compiler/src/main/kotlin/visitors/typechecker.kt

```diff
@@ -233,25 +233,60 @@ class TypeChecker(symbolTable: Table) : ScopedASTVisitor(symbolTable = symbolTab
         node.setType(node.arr.getType())
     }
 
-    override fun visit(node: ArrayBodyExpr) {
+    /**
+     * Finds the size of the largest subarray in each dimension.
+     * Note: Note if their is empty array literals in the tree, this function may return fewer dimensions than expected.
+     */
+    private fun searchSize(node: ArrayLiteralExpr, sizes: MutableList<Int> = mutableListOf(), depth: Int = 1): MutableList<Int> {
+        if (sizes.size < depth) {
+            sizes.add(node.values.size)
+        } else  if (sizes[depth-1] < node.values.size) {
+            sizes[depth-1] = node.values.size
+        }
+
+        @Suppress("NAME_SHADOWING") var sizes = sizes
+        node.values.forEach {
+            if (it is ArrayLiteralExpr) {
+                sizes = searchSize(it, sizes, depth + 1)
+            }
+        }
+
+        return sizes
+    }
+
+    private fun pushDownSize(n: Expr, size: List<Int>) {
```

Comment:
Needs a javadoc

Reply:
```suggestion
    /**
     * Get the size declared as part of the sized arrays type declaration.
     *
     * @return Returns a list of sizes from the type declaration, and null for any dimension which has to be inferred. Example "[10][][30]bool{}" would return 10, null, 30
     */
    private fun pushDownSize(n: Expr, size: List<Int>) {
```

Reply:
That's the same javadoc as `getArrayDeclaredSize()` ??

Reply:
```suggestion
    /**
     * Updates an ArrayLiteral subtree with sizes.
     * Fill the size field of ArrayLiteral according to it's place in the tree and the values from size.
     */
    private fun pushDownSize(n: Expr, size: List<Int>) {
```
Sorry I copied the wrong javadoc